File: 1.Stack/Cpp/Stack.cpp
#include <iostream>
#define MAX 5

struct Stack
{
    int TOS;       // Top of Stack index
    int DATA[MAX]; // Array to store stack elements
};

typedef struct Stack St;

// Is Empty
int isEmpty(St *myStack)
{
    return myStack->TOS == -1; // 1 if empty, 0 otherwise
}

// Is Full
int isFull(St *myStack)
{
    return myStack->TOS == MAX - 1; // 1 if full, 0 otherwise
}

// Push
void push(St *myStack, int element)
{
    if (isFull(myStack))
    {
        std::cout << "Stack is Full!" << std::endl;
    }
    else
    {
        myStack->TOS++;
        myStack->DATA[myStack->TOS] = element;
        std::cout << element << " pushed to the stack!" << std::endl;
    }
}

// Peek
int peek(St *myStack)
{
    if (isEmpty(myStack))
    {
        std::cout << "The Stack is Empty!" << std::endl;
        return -1; // Error value
    }
    else
    {
        return myStack->DATA[myStack->TOS];
    }
}

// Pop
int pop(St *myStack)
{
    if (isEmpty(myStack))
    {
        std::cout << "The Stack is Empty!" << std::endl;
        return -1; // Error value
    }
    else
    {
        int element = myStack->DATA[myStack->TOS];
        myStack->TOS--; // Move TOS down
        std::cout << element << " popped from the stack!" << std::endl;
        return element;
    }
}

int main()
{
    // Initializing a stack
    St myStack;
    myStack.TOS = -1; // Initialize TOS

    push(&myStack, 4);
    push(&myStack, 3);
    push(&myStack, 2);

    std::cout << "Top Element: " << peek(&myStack) << std::endl;

    pop(&myStack);
    std::cout << "Top Element: " << peek(&myStack) << std::endl;

    pop(&myStack);
    std::cout << "Top Element: " << peek(&myStack) << std::endl;

    return 0;
}

File: 1.Stack/Python/Stack.py
# Stack Implementation in Python Using Lists
class Stack:
    def __init__(self):
        self.items = []  # Initialize an empty list to store stack elements
    
    # Method to add an element to the stack
    def push(self, item):
        self.items.append(item)  # Append the item to the end of the list
    
    # Method to remove and return the top element of the stack
    def pop(self):
        if not self.is_empty():  # Check if the stack is not empty
            return self.items.pop()  # Remove and return the last element
        raise IndexError("pop from empty stack")  # Raise an error if the stack is empty

    # Method to view the top element without removing it
    def peek(self):
        if not self.is_empty():  # Check if the stack is not empty
            return self.items[-1]  # Return the last element without removing it
        raise IndexError("peek from empty stack")  # Raise an error if the stack is empty

    # Method to check if the stack is empty
    def is_empty(self):
        return len(self.items) == 0  # Return True if the list is empty, False otherwise

    # Method to get the size of the stack
    def size(self):
        return len(self.items)  # Return the number of elements in the stack

# Example Usage
stack = Stack()  # Create an instance of the Stack class
stack.push(1)    # Push 1 onto the stack
stack.push(2)    # Push 2 onto the stack

print(stack.peek())  # Output: 2 (Top element of the stack)
print(stack.pop())   # Output: 2 (Remove and return the top element)
print(stack.is_empty())  # Output: False (The stack still has elements)
print(stack.size())  # Output: 1 (Size of the stack after popping one element)
